Update support/example for ESP32-C3 to use latest versions of dependencies

jessebraham opened the pull request:
This updates the `rtic` and `rtic-macros` packages to handle some changes which have been made since support for the ESP32-C3 has been updated last. It additionally updates the `sw_and_hw` example for this device, which I have tested and verified using a `ESP32-C3-DevKit-RUST-1 v1.2a` board.

If there are any questions or concerns please let me know, I'm happy to make any requested changes.

This supersedes #973 as well, so that can be closed when/if this is merged.

jessebraham committed:
Update `rtic` package to use latest version of `esp32c3` dependency

jessebraham committed:
Update `rtic-macros` ESP32-C3 bindings to reflect changes in HAL

jessebraham committed:
Update the ESP32-C3 examples to use latest versions of all dependencies

jessebraham commented on examples/esp32c3/examples/sw_and_hw.rs:
```diff
@@ -5,32 +5,32 @@
         let peripherals = Peripherals::take();
-        let io = IO::new(peripherals.GPIO, peripherals.IO_MUX);
-        let mut button = io.pins.gpio9.into_pull_down_input();
+        let io = Io::new_no_bind_interrupt(peripherals.GPIO, peripherals.IO_MUX);
```
I will note that this is a bit of a rough edge for which we don't have a good solution yet. Essentially the issue is that `Io::new()` binds a handler to the `GPIO` interrupt for async functionality, so if we want to use this interrupt for other purposes (like in the `sw_and_hw` example, when creating the hardware task) then we need to _not_ bind the interrupt in the constructor.

This is potentially a bit of a footgun, so I can add a comment here if necessary.

jessebraham replied:
There is a tracking issue for this in the `esp-hal` repo, FYI: https://github.com/esp-rs/esp-hal/issues/2009

jessebraham committed:
Update changelogs

jessebraham commented:
I'm not sure what you'd like me to do about the QEMU failure, I would not expect this output to stay consistent (as we're seeing with this failure).

onsdagens commented on rtic/src/export/riscv_esp32c3.rs:
```diff
@@ -138,28 +138,30 @@ pub fn unpend(int: Interrupt) {
+    const INTERRUPT_MAP_BASE: u32 = 0x600C_2000;
+    const RESERVED_INTERRUPTS: &[u8] = &[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15];
+
+    if RESERVED_INTERRUPTS.contains(&cpu_int_id) {
```
The line [here](https://github.com/rtic-rs/rtic/pull/975/files#diff-3fe7541514399a332f826c2d8567f0708a53cd71bcadd777705979177f0b2beeR229) already ensures this doesn't happen. 

This function is not meant to be called directly by the user, and is side-effectful, maybe it should be marked as `unsafe` to reflect that?

onsdagens commented on rtic/src/export/riscv_esp32c3.rs:
```diff
@@ -138,28 +138,30 @@ pub fn unpend(int: Interrupt) {
+    if RESERVED_INTERRUPTS.contains(&cpu_int_id) {
+        panic!("interrupt {cpu_int_id} is reserved!");
+    }
+    if prio == 0 {
```
Similarly, no interrupt will ever actually be configured at priority 0 by RTIC. We can add an additional compile-time check to really make sure.

onsdagens committed:
adjust expected qemu output, add compile-time checks

onsdagens committed:
remove runtime checks, this is checked at compile time

onsdagens commented:
Tested this locally on the same board, and it seems to work as expected.

I've fixed up the QEMU expected output, applied my suggestions and opened a PR on your fork.

onsdagens committed:
fix expected qemu output

onsdagens commented:
i see my review is set as pending, so i'm not sure if you can see my comments. in any case, i see now what you were saying about QEMU, the output does not match what is expected, and what actually happens when running on a board

i think we should keep QEMU around, adjust the expected output to the unexpected results and use it as a marker for when something is potentially wrong and needs to be run on a board.

jessebraham commented:
I agree, it's probably worth keeping the checks around, however perhaps being a bit less strict with regards to the bootloader output would be best.

Anyway, with your changes merged CI is green here now at least, so thanks for that!

jessebraham committed:
Clean up interrupt enable code a bit

onsdagens commented:
i guess i just forgot to press submit, the comments are now outdated but i will leave them here for the record.

korken89 commented:
Great work, thank you!

I just merged some changes from @onsdagens , could you double check this PR is still OK?
E.g. there was a reference to `esp32c3` v0.22 in `rtic-montonics` now as well.

jessebraham commented:
Thanks for taking a look. I have rebased my branch locally but am running into some issues with the monotonic example; will investigate further.

Might take me a few days to get to this, lots going on right now for me :)

korken89 commented:
Thank you, no rush!

jessebraham committed:
Update `rtic-monotonic` to use the latest PAC for ESP32-C3

jessebraham committed:
Update `CHANGELOG.md` for `rtic-monotonic`

jessebraham commented:
I've resolved the issues I had encountered with the monotonic example, seems to be working correctly now.

Unfortunately, again we are having issues with the checks on the QEMU output. Happy to make whatever changes here, but I am not very familiar with these checks so am not sure how best to resolve this long-term. I think that the bootloader output can probably be ignored entirely in the QEMU checks.

AfoHT commented:
I agree z

> I've resolved the issues I had encountered with the monotonic example, seems to be working correctly now.
> 
Awesome! :rocket: 
> Unfortunately, again we are having issues with the checks on the QEMU output. Happy to make whatever changes here, but I am not very familiar with these checks so am not sure how best to resolve this long-term. I think that the bootloader output can probably be ignored entirely in the QEMU checks.

I agree, bootloader is not the main part. Do you want to attempt implement the ignoring of bootloader output, or should we join forces on that one?

jessebraham commented:
> Do you want to attempt implement the ignoring of bootloader output, or should we join forces on that one?

Would be great if I could get a hand with that, please. New project at work will be consuming a fair bit of my brainpower and time for at least the next couple weeks here.

AfoHT committed:
ci: esp32c3: Format runner.sh

AfoHT committed:
ci: esp32c3: Default to silent boot

export DEBUGGING while running to get verbose boot

env DEBUGGING=1 cargo xtask ...

AfoHT committed:
ci: esp32c3: Update expected example output

AfoHT reviewed (approved):
:rocket:

jessebraham commented:
Thank you for wrapping this up!
